_posts/2019-12-31-memkind-dax-kmem.md

```diff
@@ -0,0 +1,215 @@
+---
+title: Memkind support for KMEM DAX option
+author: michalbiesek
+layout: post
+identifier: memkind-support-for-KMEM-DAX-option
+---
+
+### Introduction
+
+Linux kernel version 5.1 brings in [support][kernel-patch] for the volatile-use of persistent memory
+as a hotplugged memory region (**KMEM DAX**).
+From the perspective of OS, persistent memory will be seen as a separate NUMA node(s).
+[libmemkind][memkind-release] extends its API by new kinds which allow for automatic detecting and allocating to
+persistent memory.
+
+### Requirements
+
+At least kernel 5.1 version with enabled **KMEM DAX** option
+
+{% highlight console %}
+$ make nconfig
+	Device Drivers --->
+		-*- DAX: direct access to differentiated memory --->
+			<M>   Device DAX: direct access mapping device
+			<M>     PMEM DAX: direct access to persistent memory
+			<M>     KMEM DAX: volatile-use of persistent memory
+			< >   PMEM DAX: support the deprecated /sys/class/dax interface
+{% endhighlight %}
+
+[ndctl and daxctl][ndctl-release] with version 66 or later.
+
+### System Configuration
+
+Using NVDIMM as an extension of DRAM needs proper configuration with `ndctl` and `daxctl`.
+A Reconfiguration of Device-DAX depends on dax-bus device model.
+Kernel should support `/sys/bus/dax model`. To migrate it from
+`sys/class/dax` to `/sys/bus/dax model` please use [daxctl-migrate-device-mode][daxctl-migrate-device-model].
+
+### Reconfigure dax devices
+
+To see the list of available NUMA nodes on the system you could use `numactl`.
+An example of initial configuration is presented below:
+
+{% highlight sh %}
+$ numactl -H
+
+available: 2 nodes (0-1)
+node 0 cpus: 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83
+node 0 size: 80249 MB
+node 0 free: 68309 MB
+node 1 cpus: 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111
+node 1 size: 80608 MB
+node 1 free: 71958 MB
+node distances:
+node   0   1
+  0:  10  21
+  1:  21  10
+{% endhighlight %}
+
+To [create a namespace][ndctl-create-namespace] in Device-DAX mode as standard memory from all the available capacity of NVDIMM:
+
+{% highlight sh %}
+$ ndctl create-namespace --mode=devdax --map=mem
+{% endhighlight %}
+
+To [list DAX devices][daxctl-list]:
+
+{% highlight sh %}
+$ daxctl list
+
+[
+  {
+    "chardev":"dax1.0",
+    "size":539016298496,
+    "target_node":3,
+    "mode":"devdax"
+  }
+]
+{% endhighlight %}
+
+To [reconfigure][daxctl-reconfigure-device] DAX device from `devdax` mode to a `system-ram-mode`:
+
+{% highlight console %}
+$ daxctl reconfigure-device dax1.0 --mode=system-ram
+{% endhighlight %}
+
+With this operation persistent memory could be seen as separate NUMA and be used as volatile RAM.
```

Comment:
After this operation, persistent memory will be configured as separate NUMA node and can be used as volatile memory.

Reply:
Done